transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown

```diff
@@ -0,0 +1,117 @@
+
+NOTE: There are additional complications when hosting entities crash (i.e. an unintentional shutdown occurs). This extreme scenario is beyond what WildFly can handle and the hosting cloud environment must provide the ability to recover the hosting entity from crashes. In this case, the state of the hosting entity (e.g. its file system, ip address, etc.) should be recovered by the cloud environment without the intervention of the administrator. In K8s, for example, the employment of StatefulSet creates a stable environment where crashes are handled automatically. If a particular cloud environment does not provide a guarantee similar to StatefulSet, it is not possible to employ WildFly’s Transaction Subsystem. *As there are no solutions to solve this extreme scenario within WildFly (as it is impossible to control the lifecycle of the host from WildFly), this proposal is only concerned with problem 1.2.*
+
+== Issue Metadata
+
+=== Issue
+
+* [WFLY-17742](https://issues.jboss.org/browse/WFLY-17742)
+
+=== Dev Contacts
+
+* mailto:{jfinelli@redhat.com}[{Manuel Finelli}]
+
+=== QE Contacts
+
+* mailto:{istraka@redhat.com}[{Ivan Straka}]
+
+=== Testing By
+
+* [ ] Engineering
+* [X] QE
+
+=== Affected Projects or Components
+
+* [wildfly-core](https://github.com/wildfly/wildfly-core)
+* [wildfly](https://github.com/wildfly/wildfly)
+* [narayana](https://github.com/jbosstm/narayana)
+
+=== Relevant Installation Types
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+* This proposal is about importing into WildFly the transactional functionalities of its Kubernetes Operator. It does not include any further improvements
+* The Management Model of WildFly will not be modified
+* The Transactions subsystem needs to become aware of in-doubt and heuristic transactions and delay its suspension as long as all transactions are completed. This will guarantee data integrity in all scenarios, including situations when a negative timeout (i.e. indefinite waiting) is employed (e.g. in cloud environments)
+* The administrator needs to receive notifications when the Transactions subsystem is causing delays in the graceful shutdown of WildFly, achieved through integration with the existing logging mechanism. This is especially true when a negative timeout is used (i.e. indefinite graceful shutdown). Moreover, internal details like 'ServerActivity' do not need to be prominent. Part of working out such a notification would be how to consistently describe a ServerActivity to the administrator in an understandable manner without exposing internal technical implementation details
```

Comment:
WildFly Administrators can configure WildFly graceful shutdown in two different ways:

* Timeout > 0: WildFly will shut down after X seconds, even though there are ServerActivities that are still trying to shut down gracefully.
* Timeout < 0: WildFly will shut down only when all ServerActivities have completed their graceful shutdowns.

Considering the restriction we already discussed (i.e., we cannot change the Management Model of WildFly with this proposal, which means we cannot introduce a switch to turn off the graceful shutdown once it has started), using the timeout of WildFly graceful shutdown is the best option we have (i.e., inform/advice Administrators to use the timeout as a switch). The switch to dynamically stop WildFly graceful shutdown will be introduced as a further improvement once this proposal is implemented.

Reply:
Answers in-line:

> So in the `Timeout < 0` case what do we tell the admin to do when the shutdown hangs?

Documentation will be updated to explain the revised semantic of `timeout < 0`. Even though a negative timeout has always meant that WildFly was gracefully shutting down indefinitely, WFLY-17742 will add to the current meaning that the Transactions subsystem can force WildFly to delay indefinitely its graceful shutdown as long as there are transactions to be completed. As a consequence, the admin will be advised to use the usual procedure to solve in-doubt/heuristic transactions (which is already documented).

>And will the documentation advise that administrators never use a Timeout > 0 (since we can't guarantee ACID semantics for this value)?

This is already a shortcoming of the documentation. As we discussed privately, I will raise a JBEAP to document this shortcoming.

> What are the semantics of the `Timeout = 0` case?

It is already [documented](https://docs.wildfly.org/30/Admin_Guide.html#Suspend) and this proposal won't modify its semantic.

> Are other people OK about telling admins that they should debug shutdown hangs by searching through log files - as an admin I'd want something more formal so that the reaction to the event can be scripted and integrated into whatever existing management applications I am using.

This is the state of things we currently have in the [WildFly operator](https://github.com/wildfly/wildfly-operator). As reported already, this proposal is an effort to import in WildFly some functionalities of the WildFly operator. Moreover, as discussed in other places (privately), once this proposal gets implemented we will consider the introduction of a switch to force the graceful shutdown (as we cannot modify the management model of WildFly with this proposal)

Reply:
> Also while I'm here, so as not to start a new thread going and because it's related to this thread, how does the heuristic part of the proposal relate to the existing [proposal about heuristics](https://github.com/wildfly/wildfly-proposals/pull/295)? I think I may have asked this one before but I can't find a record of the answer.

As WildFly's `existing logging mechanism` will be used, we can integrate the new logs into K8s.

NOTE: I cannot resolve this comment as otherwise I will hide the entire conversation :-)

Reply:
I asked these questions because the proposal does not contain the answers to them so answering my questions directly in the PR does not help the next reader so I think the proposal must provide clarity, unless it does and I've not read it correctly.